### PR TITLE
Microapp

### PR DESCRIPTION
**Pull request by mrquincle**
Bug fixes in twi/ic2 code.
Bug fixes in microapp.
Addition of additional option field for presence/absence of newlines in microapps.
Functionality tested with MCP9808 temperature sensor.

### CODE REVIEW

File: source/conf/cmake/CMakeBuild.config.default [hunk 1]
```diff
@@ -427,9 +427,6 @@ FLASH_MICROAPP_BASE=0x69000
 # The number of pages for microapps (each page is 0x1000 / 4096 bytes)
 FLASH_MICROAPP_PAGES=2
 
-# The start of where the microapp is allowed to put its memory (start of its .bss .data and then the heap)
-RAM_MICROAPP_BASE=0x2000B000
-
-# The amount of RAM it is allowed to use (stack grows down from RAM_MICROAPP_BASE + RAM_MICROAPP_AMOUNT)
-RAM_MICROAPP_AMOUNT=0x2000
+# The amount of RAM it is allowed to use (microapp is placed before IPC ram, and grows down).
```
Comment (mrquincle) [inline source/conf/cmake/CMakeBuild.config.default:430]: The comment "and grows down" is a bit confusing. The microapp is allocated a fixed amount `RAM_MICROAPP_AMOUNT` and will use parts at the beginning and at the end of this piece of ram.

File: source/include/cfg/cs_StaticConfig.h.in [hunk 2]
```diff
@@ -18,7 +18,8 @@
 extern "C" {
 #endif
 
-const uintptr_t g_RAM_MICROAPP_END = (@RAM_MICROAPP_BASE@ + @RAM_MICROAPP_AMOUNT@);
+// Microapp ram is located before IPC ram.
+const uintptr_t g_RAM_MICROAPP_END = @RAM_R1_BASE@ + @RAM_APPLICATION_AMOUNT@ - @RAM_BLUENET_IPC_LENGTH@;
```
Comment (mrquincle) [inline source/include/cfg/cs_StaticConfig.h.in:22]: This seems wrong... I suppose the amount of ram for bluenet is `RAM_APPLICATION_AMOUNT` and this should exclude the amount of ram for the microapp. Hence, `g_RAM_MICROAPP_END` should be at:

```
const uintptr_t g_RAM_MICROAPP_END = @RAM_R1_BASE@ + @RAM_APPLICATION_AMOUNT@ + @RAM_MICROAPP_AMOUNT@;
```

Comment (mrquincle) [reply source/include/cfg/cs_StaticConfig.h.in:22]: That, or rename, `@RAM_APPLICATION_AMOUNT@` to something like `@RAM_TOTAL_AMOUNT@` (or something better).

File: source/include/third/nrf/generic_gcc_nrf52.ld [hunk 3]
```diff
@@ -33,14 +33,21 @@ INCLUDE "nrf_symbols.ld"
  *
  * Update the sections using a recent flash_placement.xml file...
  */
+
+RAM_BLUENET_IPC_START = (RAM_R1_BASE + RAM_APPLICATION_AMOUNT - RAM_BLUENET_IPC_LENGTH);
```
Comment (mrquincle) [inline source/include/third/nrf/generic_gcc_nrf52.ld:37]: Same, quite confusing. 

Actually I see `RAM_APPLICATION_AMOUNT` mainly used to define the end of the ram area which we can use and then subtracting stuff again. 

I think defining something like `RAM_R1_END` to be at the ipc start would simplify a few equations. With, again, just defining a setting that indicates total ram available. Then the developer does not have to calculate this in the config file. Moreover, if we go to different hardware this value is then very easy to change.

**Review by mrquincle — COMMENTED** [review]
Some suggestions to make this a bit more transparent to developers.

File: source/include/cfg/cs_StaticConfig.h.in (hunk 2 above)
Comment (vliedel) [reply source/include/cfg/cs_StaticConfig.h.in:22]: I left the meaning of `RAM_APPLICATION_AMOUNT` as it was, it already included the IPC ram.

**Pull request merged**